python/hail/methods/statgen.py

```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
+        - AD: The filtered alleles' columns are eliminated, e.g., filtering alleles 1 and 2 transforms ``25,5,10,20`` to ``25,20``.
+        - DP: No change.
+        - PL: The filtered alleles' columns are eliminated and the remaining columns shifted so the minimum value is 0.
+        - GQ: The second-lowest PL (after shifting).
+
+        """
+        ds = self._ds
+        newPL = functions.bind(
+            functions.cond(
+                functions.is_defined(ds.PL),
+                functions.range(0, self.new_v.num_genotypes()).map(
+                    lambda newi: functions.bind(
+                        functions.call(newi),
+                        lambda newc: ds.PL[functions.gt_index(
+                            self.new_to_old[newc.gtj()], self.new_to_old[newc.gtk()])])),
+                functions.null(TArray(TInt32()))),
+            lambda unnorm: unnorm - unnorm.min())
+        self.annotate_entries(
+            GT = functions.call(functions.gt_from_pl(newPL)),
+            AD = functions.cond(
+                functions.is_defined(ds.AD),
+                functions.range(0, self.new_v.num_alleles()).map(
+                    lambda newi: ds.AD[self.new_to_old[newi]]),
+                functions.null(TArray(TInt32()))),
+            # DP unchanged
+            GQ = functions.gq_from_pl(newPL),
+            PL = newPL)
+
+    def downcode_entries_hts(self):
+        """Use the downcode algorithm to update the matrix table entries for
+        the new, allele filtered variant.  This requires the entries
+        have the HTS genotype schema:
+
+        **The downcode algorithm**
+
+        .. code-block:: text
+
+            GT: Call,
+            AD: Array[!Int32],
+            DP: Int32,
+            GQ: Int32,
+            PL: Array[!Int32].
+
+        The downcode algorithm recodes occurances of filtered alleles
+        to occurances of the reference allele (e.g. 1 -> 0 in our
+        example). So the depths of filtered alleles in the AD field
+        are added to the depth of the reference allele. Where
+        downcodeing filtered alleles merges distinct genotypes, the
+        minimum PL is used (since PL is on a log scale, this roughly
+        corresponds to adding probabilities). The PLs are then
+        re-normalized (shifted) so that the most likely genotype has a
+        PL of 0, and GT is set to this genotype.  If an allele is
+        filtered, this algorithm acts similarly to
+        :meth:`~hail.VariantDataset.split_multi`.
+
+        The downcoding algorithm would produce the following:
+
+        .. code-block:: text
+
+          GT: 0/1
+          GQ: 10
+          AD: 35,50
+
+          0 | 20
+          1 | 0    10
+            +-----------
+              0    1
+
+        In summary:
+
+        - GT: Downcode filtered alleles to reference.
+        - AD: The filtered alleles' columns are eliminated and their
+          value is added to the reference, e.g., filtering alleles 1
```

Comment:
"Columns of filtered alleles are eliminated and their values are added to the reference column"